Add guidance for trusted publishing

chrysle opened the pull request:
See https://github.com/pypa/packaging.python.org/pull/1251#issuecomment-1555881672 for context.

##### Preview

https://python-packaging-user-guide--1261.org.readthedocs.build/en/1261/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -51,6 +51,31 @@ Let's begin! 🚀
       create it. It's not the same as a regular PyPI account.
 
 
+Using trusted publishing
```
The idea of the guide is to be a simple and straightforward starting point. It should demonstrate a flow without any deviations, it should be linear.
Also, OIDC is more important as the default for anybody using GitHub. We can mention that it's possible to use login+password when publishing to third party indexes but it mustn't be the central part of this document.

chrysle replied:
I agree and have adapted the documentation and release workflow file.

webknjaz replied:
Marking this thread as resolved.

webknjaz reviewed (commented):
Let's not add it as a second-class citizen but instead replace it with OIDC entirely.

chrysle commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -150,3 +175,4 @@ sure that your release pipeline remains healthy!
    https://github.com/marketplace/actions/pypi-publish
 .. _Secrets:
    https://docs.github.com/en/actions/reference/encrypted-secrets
+.. _trusted publishing: https://docs.pypi.org/trusted-publishers/
```
The PyPI docs haven't configured that. They'd need the [mkdocstrings](https://mkdocstrings.github.io/usage/#cross-references-to-other-projects-inventories) plugin.

I consider adding it there.

webknjaz replied:
I'm not sure. This doc describes referencing other inventories, not producing one..

chrysle replied:
See the last example.

webknjaz replied:
Oh, I see now. Okay.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -6,6 +6,8 @@ jobs:
   build-n-publish:
     name: Build and publish Python 🐍 distributions 📦 to PyPI and TestPyPI
     runs-on: ubuntu-latest
+    permissions:
```
Since this adds extra lines, please verify that it doesn't break included fragments in RST, especially line highlights.

webknjaz replied:
It's strongly recommend to use environments to further lock it down:
```suggestion
    environment:
      name: pypi
      url: https://pypi.org/p/<package-name>
    permissions:
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -31,10 +33,7 @@ jobs:
     - name: Publish distribution 📦 to Test PyPI
```
I've been actively discouraging publishing from the same job as building and publishing to PyPI and TestPyPI. We need to rewrite this workflow to split them. The latter would have `testpypi` as the environment name, allowing people to use differing protections on the GH side + the URL in the UI would be different as well. Also, the job building dists mustn't have access to `id-token` since it may potentially run untrusted code.

webknjaz replied:
@chrysle I think you might've missed this comment.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +17,43 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+to `GitHub Actions CI/CD`_. This is recommended for security reasons, since 
+the generated tokens are created for each of your projects
+individually and expire automatically. Otherwise you'll need to generate an
+`API token`_ or provide a username/password combination for both PyPI and
```
I wouldn't suggest using username and password for PyPI or TestPyPI. Instead, I'd only mention such a possibility in the context of publishing to third party indexes like devpi.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +17,43 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
-settings.
+Since this guide will demonstrate uploading to both
+PyPI and TestPyPI, we'll need two trusted publishers configured. 
+The following steps will lead you through creating the "pending" publishers.
```
Let's mention that it's also possible to configure trust for pre-existing projects, if one is an owner.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +17,43 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
-4. Now, go to https://test.pypi.org/manage/account/#api-tokens
-   and repeat the steps. Save that TestPyPI token on GitHub
-   as ``TEST_PYPI_API_TOKEN``.
+1. Go to https://pypi.org/manage/account/publishing/
```
We probably need a trailing period for style consistency here:
```suggestion
1. Go to https://pypi.org/manage/account/publishing/.
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +17,43 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
-   as ``TEST_PYPI_API_TOKEN``.
+1. Go to https://pypi.org/manage/account/publishing/
+2. Fill in the name you wish to publish your new project under,
+   your repository data and the name of the release workflow file 
```
Since it's a tutorial, we shouldn't send people scroll the page back and forth. Instead, we have specific values in the examples and should explicitly tell the reader what to enter.

webknjaz reviewed (commented):
Here's some initial feedback. But I might've missed something reviewing from the phone..

chrysle commented:
Thanks! I tried to apply your review feedback best as possible.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -27,14 +32,40 @@ jobs:
+        pip install
+        build
+        --user
+    - name: Build a binary wheel and a source tarball
```
As I wrote in https://github.com/pypa/packaging.python.org/pull/1261/files#r1208671692, building should happen in a separate job, not each job. The publishing jobs **must not** have a build step at all.

I've got an example of the publishing jobs showing how to get dists out of the GHA artifacts at https://github.com/sphinx-contrib/sphinxcontrib-towncrier/blob/098c781/.github/workflows/ci-cd.yml#L623-L682. There's a build job there too, but it wraps commands with tox. Here, we'll want to keep the pypa/build invocation, just look up how to store the artifacts, maybe.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -27,14 +32,40 @@ jobs:
+      run: >-
+        python3 -m
+        build
+        --sdist
+        --wheel
+        --outdir dist/
+        .
```
When moving the build step, plz drop all these arguments. They are not needed in the simple case this document describes. Also, declaring `--sdist --wheel` on CLI can be harmful in some cases.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -3,10 +3,10 @@ name: Publish Python 🐍 distributions 📦 to PyPI and TestPyPI
+    name: Build the source package
     runs-on: ubuntu-latest
-
+    
```
```suggestion

```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +23,44 @@ jobs:
       with:
-        password: ${{ secrets.TEST_PYPI_API_TOKEN }}
-        repository-url: https://test.pypi.org/legacy/
+        name: python-package-distributions     
```
```suggestion
        name: python-package-distributions
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +23,44 @@ jobs:
-        password: ${{ secrets.TEST_PYPI_API_TOKEN }}
-        repository-url: https://test.pypi.org/legacy/
+        name: python-package-distributions     
+  build-n-publish-pypi:
```
```suggestion
  publish-to-pypi:
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +23,44 @@ jobs:
     - name: Publish distribution 📦 to PyPI
       if: startsWith(github.ref, 'refs/tags')
       uses: pypa/gh-action-pypi-publish@release/v1
+  build-n-publish-testpypi:
```
```suggestion
  publish-to-testpypi:
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +23,44 @@ jobs:
+        name: python-package-distributions
+        path: dist/
     - name: Publish distribution 📦 to PyPI
       if: startsWith(github.ref, 'refs/tags')
```
Such conditionals should be moved from the step level to the job level. No need to spawn an entire VM only to do nothing useful later on.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -8,7 +8,9 @@ popular choice is having a workflow that's triggered by a
 whenever a tagged commit is pushed.
-It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+It will use the `pypa/gh-action-pypi-publish GitHub Action`_ for
+publishing and `upload-artifact`_ and `download-artifact`_ actions 
```
That's a long sentence. Let's try splitting it.
```suggestion
publishing. It also uses `upload-artifact`_ and `download-artifact`_ actions 
```

chrysle commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,47 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
-   as ``TEST_PYPI_API_TOKEN``.
+1. Go to https://pypi.org/manage/account/publishing/.
+2. Fill in the name you wish to publish your new project under,
+   your GitHub username and repository name and 
+   the name of the release workflow file under 
+   the ``.github/`` folder, see :ref:`workflow-definition`. 
```
I only have [Anaconda](http://damnwidget.github.io/anaconda/) in use, that doesn't care about trailing spaces in a RST file... Can you suggest me anything?

webknjaz replied:
I think we should integrate pre-commit and use https://github.com/pre-commit/pre-commit-hooks#trailing-whitespace... I'm not sure about other projects. I know that Vim/NeoVim can be configured to show trailing spaces visually, there's probably some plugins for that. Most projects just use pre-commit for this sort of linters/formatters.

chrysle replied:
Thanks for suggesting, I've opened #1268 for that.

webknjaz commented:
Hm.. looks like I have a pending review that I didn't submit...

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
       with:
-        password: ${{ secrets.TEST_PYPI_API_TOKEN }}
-        repository-url: https://test.pypi.org/legacy/
+        name: python-package-distributions
```
We have to specify the path to the artifacts we want to upload:
```suggestion
        name: python-package-distributions
        path: dist/
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
     - name: Publish distribution 📦 to PyPI
-      if: startsWith(github.ref, 'refs/tags')
+      uses: pypa/gh-action-pypi-publish@release/v1
+  publish-to-testpypi:
```
```suggestion

  publish-to-testpypi:
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
-        password: ${{ secrets.TEST_PYPI_API_TOKEN }}
-        repository-url: https://test.pypi.org/legacy/
+        name: python-package-distributions
+  publish-to-pypi:
```
let's have visual separators between the job definitions
```suggestion

  publish-to-pypi:
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
-        repository-url: https://test.pypi.org/legacy/
+        name: python-package-distributions
+  publish-to-pypi:
+    name: Build and publish Python 🐍 distributions 📦 to PyPI
```
This job shouldn't duplicate building, it should just publish.
```suggestion
    name: >-
      Publish Python 🐍 distributions 📦 to PyPI
      and sign them with Sigstore
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
-      if: startsWith(github.ref, 'refs/tags')
+      uses: pypa/gh-action-pypi-publish@release/v1
+  publish-to-testpypi:
+    name: Build and publish Python 🐍 distributions 📦 to TestPyPI
```
```suggestion
    name: Publish Python 🐍 distributions 📦 to TestPyPI
```

sinoroc replied:
Publish distribution packages to TestPyPI

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
+      with:
+        name: python-package-distributions
+        path: dist/
+    - name: Publish distribution 📦 to Test PyPI
```
```suggestion
    - name: Publish distribution 📦 to TestPyPI
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
+        name: python-package-distributions
+  publish-to-pypi:
+    name: Build and publish Python 🐍 distributions 📦 to PyPI
+    runs-on: ubuntu-latest
```
Publishing has to wait until build produces the dists:
```suggestion
    needs:
    - build
    runs-on: ubuntu-latest
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
+      uses: pypa/gh-action-pypi-publish@release/v1
+  publish-to-testpypi:
+    name: Build and publish Python 🐍 distributions 📦 to TestPyPI
+    runs-on: ubuntu-latest
```
```suggestion
    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
    needs:
    - build
    runs-on: ubuntu-latest
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +77,77 @@ should make GitHub run this workflow:
-you use the latest release tag; a tool like GitHub's dependabot can keep
-these updated regularly.
+This step uses the `pypa/gh-action-pypi-publish`_ GitHub
+Action: After the stored distribution package has been 
```
```suggestion
Action: after the stored distribution package has been 
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -23,18 +25,43 @@ jobs:
+        path: dist/
     - name: Publish distribution 📦 to PyPI
-      if: startsWith(github.ref, 'refs/tags')
+      uses: pypa/gh-action-pypi-publish@release/v1
```
I figured, while we're on it, let's show people how to sign the dists. Especially, as GPG support is being phased out o PyPI.
```suggestion
      uses: pypa/gh-action-pypi-publish@release/v1
    - name: Sign the dists with Sigstore
      uses: sigstore/gh-action-sigstore-python@v1.2.3
      with:
        inputs: >-
          ./dist/*.tar.gz
          ./dist/*.whl
```
Only add this to one job since we don't need to sign the same artifacts twice. We don't do this in the build job since it doesn't have the OIDC privilege.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +77,77 @@ should make GitHub run this workflow:
+This step uses the `pypa/gh-action-pypi-publish`_ GitHub
+Action: After the stored distribution package has been 
+downloaded by the `download-artifact`_ action, it uploads 
+the contents of the ``dist/`` folder into PyPI unconditionally.
```
```suggestion
the contents of the ``dist/`` folder into PyPI unconditionally.
This job also signs the artifacts with Sigstore right after publishing them to PyPI.
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -2,11 +2,13 @@ name: Publish Python 🐍 distributions 📦 to PyPI and TestPyPI
 
 on: push
 
+# Only trigger this for tag changes.
+if: startsWith(github.ref, 'refs/tags/')
+
```
We can't have it on the workflow level since it'd contradict the statement we have at the very bottom, in the summary paragraph, that the dists to TestPyPI are published on non-tagged pushes.
```suggestion
```

chrysle commented:
Thanks a lot for the extensive review! Diving into this tomorrow.

sinoroc commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,48 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
-   and repeat the steps. Save that TestPyPI token on GitHub
-   as ``TEST_PYPI_API_TOKEN``.
+1. Go to https://pypi.org/manage/account/publishing/.
+2. Fill in the name you wish to publish your new project under,
```
Name of project? Name of distribution package?

* https://packaging.python.org/en/latest/glossary/#term-Project
* https://packaging.python.org/en/latest/glossary/#term-Distribution-Package

If using the terms as defined in glossary (with link) then it should be unambiguous.

webknjaz replied:
I'd still like that extra context so how about both a reference and a hint?
```suggestion
2. Fill in the name you wish to publish your new :term:`PyPI project <Project>` under (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),
```

sinoroc commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,48 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+Since this guide will demonstrate uploading to both
+PyPI and TestPyPI, we'll need two trusted publishers configured. 
+The following steps will lead you through creating the "pending" publishers
+for your new project. However it is also possible to add `trusted publishing`_
```
Maybe we should use project with link to glossary: https://packaging.python.org/en/latest/glossary/#term-Project

chrysle replied:
Can we use a relative link for that anchor?

webknjaz replied:
Usually, glossary terms are linkable through the `:term:` role.
And I see it's available indeed: https://webknjaz.github.io/intersphinx-untangled/packaging.python.org/.

webknjaz replied:
I'd still like that extra context so how about both a reference and a hint?
```suggestion
2. Fill in the name you wish to publish your new :term:`PyPI project <Project>` under (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),
```

sinoroc commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -2,11 +2,13 @@ name: Publish Python 🐍 distributions 📦 to PyPI and TestPyPI
-  build-n-publish:
-    name: Build and publish Python 🐍 distributions 📦 to PyPI and TestPyPI
+  build:
+    name: Build the source package
```
Build distribution packages (source distribution and wheel)

chrysle commented:
Updated!

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,60 @@ jobs:
     - name: Publish distribution 📦 to PyPI
-      if: startsWith(github.ref, 'refs/tags')
       uses: pypa/gh-action-pypi-publish@release/v1
+    - name: Sign the dists with Sigstore
```
I just realized that this is not enough. It doesn't publish the signatures.
This is another step that is needed:
```yaml
    - name: Upload artifact signatures to GitHub Release
      # Confusingly, this action also supports updating releases, not
      # just creating them. This is what we want here, since we've manually
      # created the release above.
      uses: softprops/action-gh-release@v1
      with:
        # dist/ contains the built packages, which smoketest-artifacts/
        # contains the signatures and certificates.
        files: dist/**
```

But since it turns into a multi-step job, I propose moving signing+uploading signatures into a separate job called `github-release`.

webknjaz replied:
No problem! You're right, it might be getting complicated, but I'm trying to still keep it small. I was hoping that by providing people with something to copy-and-paste is going to cancel out most of the complexity. Though, this means that whatever we end up publishing shouldn't have silly mistakes, as it'd confuse newbies even more.
Honestly, I'd love to come up with a reusable workflow that people could make use of, but due to the limitations of the current implementation, it's not possible to use trusted publishing in such an environment. So maybe later.

webknjaz replied:
@chrysle did you think of getting back to this? Do you need me to take over?

webknjaz reviewed (commented):
I've had this in draft for a while, and it's probably incomplete, but I'm posting the partial review to keep this going. It's better than blocking for a long time. I'll try to review again later, after the next batch of updates...

chrysle commented:
Sure, thanks for the ping! I'll send in a patch shortly.

chrysle commented:
Will get back to this tomorrow!

miketheman commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,66 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+2. Fill in the name you wish to publish your new
+   :term:`PyPI project <Project>` under
+   (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),
+   your GitHub username and repository name and
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**suggestion:**  Since users may add publishing from an org repository, telling them to add their own username may be misleading.
‏
```suggestion
   the GitHub repository owner's name and repository name and
```

webknjaz replied:
How about adding `(org or user)` clarification?

miketheman commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,68 +95,81 @@ should make GitHub run this workflow:
-   :start-after: on:
-   :end-before: steps:
-
+This will also ensure that the release workflow is only triggered
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏the updated comment states that this will only work on tagged, but the current yml example shows `on: push` - is that correct?

A later example shows `if: startsWith(github.ref, 'refs/tags/')` - that's only relevant to the testpypi workflow, not the pypi one.

Would it be better expressed as something like:

```yaml
on:
  push:
    tags:
      - '*'
```

webknjaz replied:
No, this is an incorrect assumption — the workflow can run on normal pushes and the build step would be a smoke test while the publishing would be skipped. Honestly, I'd rather refactor this to trigger on `workflow_dispatch:` but that would be too complex — maybe it's better to have a separate guide for a more complicated / comprehensive scenario.
I usually expect this to be a part of the same main CI/CD workflow that also includes tests. This way, you can also test those built artifacts before publishing.

miketheman replied:
Okay - I'll rely on your experience here. I think I've seen the build/publish steps as part of a distinct workflow, which would trigger on tags, so if the intent to keep this all in one workflow, then the refs/tags approach is fine!

webknjaz replied:
> Okay - I'll rely on your experience here. I think I've seen the build/publish steps as part of a distinct workflow, which would trigger on tags, so if the intent to keep this all in one workflow, then the refs/tags approach is fine!

Yeah, I've seen different workflows too. I just didn't want to change too many things in this PR. But later, I'd definitely rethink this. Though, I was considering a separate more in-depth document rather then extending this with multiple branches.

webknjaz replied:
Marking this as resolved.

miketheman commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,68 +95,81 @@ should make GitHub run this workflow:
 .. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
    :language: yaml
-   :start-after: runs-on:
+   :start-after: jobs:
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏this change makes `jobs:` never appear in the rendered document, so there's no place for the user to know that there needs to be a `jobs:` key.

chrysle replied:
Thanks, maybe the `start-before` directive is better here.

miketheman commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,68 +95,81 @@ should make GitHub run this workflow:
+which is mandatory that the ``pypi-publish`` actions needs to
+implement secretless trusted publishing to PyPI.
+
+.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
+   :language: yaml
+   :start-after: path: dist/
+   :end-before: steps:
```
This step shows the user the example `url: https://pypi.org/p/<package-name>` but makes no mention of which name to place in `package-name`

miketheman reviewed (commented):
Very cool to see this update! I added some comments inline, mostly in the style of https://conventionalcomments.org/ . Let me know if any are unclear or could use more details!

webknjaz commented:
In https://github.com/python-jsonschema/check-jsonschema/commit/a9a3504ef9bcd19d410a0b9f5ac25a1f184e496e#r125535746 @sirosen suggested that it may be a good idea to have the entire workflow included in the document.
I think we can make another snippet inclusion with the entire YAML file, probably wrapped as a collapsible block.

chrysle commented:
> I think we can make another snippet inclusion with the entire YAML file, probably wrapped as a collapsible block.

Done, I hope the `sphinx-toolbox` dependency is fine.

sirosen commented:
Looking at the preview to see the changes; I have a few notes.

First, thanks for adding the complete example! It looks great to me.

Is there some explanation of sigstore which could be linked? Even as someone who tries to stay on top of things and reads the python discourse, this one isn't familiar to me. Which suggests that most readers of this guide won't know what it is. Plus, with pypi phasing out gpg signatures, anything that sounds like "signing a package" probably deserves some extra explanation.

I see the guide recommends using `softprops/action-gh-release`. I only looked into it minimally when I learned that github had deprecated their release artifact action, but some of that repo made our team a bit wary. (In particular, the github release stream vs the `v1` tag sends very mixed messages.)
I've been tinkering with using the `gh` CLI to do this instead, which should be something like
```yaml
      - name: Publish to GitHub Release
        env:
          GH_TOKEN: ${{ github.token }}
          GH_REPO: ${{ github.repository }}
        run: gh release upload "${{ github.ref_name }}" dist/*
```
(Caveat emptor: I think that's right, but I haven't tested it yet.)
If I get this working, would it be interesting to put into the guide instead, or no? I can make sure to share back the results of my tinkering/testing when it's done if so.

Last thing: I've used `sphinx-design` and been pretty happy with it. `sphinx-toolbox` looks to be similarly popular, but if there's any issue with `sphinx-toolbox`, there's another nice alternative. :slightly_smiling_face:

sethmlarson commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,99 @@ should make GitHub run this workflow:
+the contents of the ``dist/`` folder into PyPI unconditionally.
 
-.. tip::
+Signing the distribution packages
```
Do we want to have Sigstore signing in the guide for publishing with Trusted Publishers? These are different concepts and the current reading makes it seem like you need both to accomplish the task?

webknjaz replied:
@sethmlarson this is not a tutorial about trusted publishing but about publishing to PyPI from GHA which happens to use OIDC instead of long-lived credentials now. Though, I get your point and it's probably hard to pick where to draw the line between what's in and out of the scope. Perhaps, a `.. note::` admonition could explain that this job is not mandatory.
But I do feel like it's important to show people a friction-free way of signing their artifacts, especially given the now-official deprecation of GPGs in Warehouse.

chrysle commented:
> Is there some explanation of sigstore which could be linked? Even as someone who tries to stay on top of things and reads the python discourse, this one isn't familiar to me. Which suggests that most readers of this guide won't know what it is. Plus, with pypi phasing out gpg signatures, anything that sounds like "signing a package" probably deserves some extra explanation.

I already linked the [marketplace site ](https://github.com/marketplace/actions/gh-action-sigstore-python) of the `sigstore-python` GitHub Action (which originally is a command line tool) which contains all the documentation I have found.

> I see the guide recommends using `softprops/action-gh-release`. I only looked into it minimally when I learned that github had deprecated their release artifact action, but some of that repo made our team a bit wary. (In particular, the github release stream vs the `v1` tag sends very mixed messages.) 

Could you elaborate on this, probably provide an example? Maybe it's also a good idea to raise an issue.

sirosen commented:
> I already linked the [marketplace site ](https://github.com/marketplace/actions/gh-action-sigstore-python) of the `sigstore-python` GitHub Action (which originally is a command line tool) which contains all the documentation I have found.

Maybe I'm just out of the loop on this, but this is my first encounter with Sigstore. The result is that even with a link to the GitHub Action docs, I was missing all of the important context of "what is Sigstore?", "what value does signing with Sigstore provide my users?", "is this different from GPG signed packages?", etc.

I'm not sure exactly what linkage or explanation would have been best.
Maybe a link going straight to https://docs.sigstore.dev/ is the right place?

Mostly, as an outsider providing feedback, I wanted to make sure there's no assumption that readers have any prior notion of what Sigstore is or how to use it. Telling users to use the action, but without the context to know what it does, naturally raises these questions.

I guess you could see this as a fault of the sigstore action docs? What matters most, IMO, is that as a reader is that I get relatively quickly to some place where I know whether or not I can or should omit that part of the workflow.

> > I see the guide recommends using `softprops/action-gh-release`. I only looked into it minimally when I learned that github had deprecated their release artifact action, but some of that repo made our team a bit wary. (In particular, the github release stream vs the `v1` tag sends very mixed messages.)
> 
> Could you elaborate on this, probably provide an example? Maybe it's also a good idea to raise an issue.

There were basically two things that made us slightly uncomfortable.
- a growing backlog, including a backlog of PRs, with infrequent recent touches from the repo owner[^1]
- a "v1" tag, but [releases](https://github.com/softprops/action-gh-release/releases) numbered under 0.1.x

The versioning is strange and is what I referred to as "mixed messages" -- is this a "v1 stable" project, or a "v0.x unstable" project?

[^1]: I'm guilty of letting PRs lag on some projects too, so I understand how that can happen. But it often is a sign of burnout or disinvestment in the project by the author, and it can become a feed-forward process. It might be best to try to find a new maintainer, but that's a difficult process in its own right. None of this is meant as criticism. It's an assessment of the maintenance and health of the project.

Looking for another path for this functionality, the GitHub CLI seems more appropriate because GitHub and their (large, non-volunteer) team is responsible for making sure that `gh release upload` works.

The fact that GitHub has dropped their own action (https://github.com/actions/upload-release-asset) and then pointed at someone else's version of the same looks strange to me. There's a sympathetic reading of this -- that they wanted to boost someone else's project -- but I have a more critical take. It's not directly relevant to assessing the long-term stability of these tools though, so I'll refrain from wandering too far OT.

webknjaz commented:
@chrysle looks like https://github.com/pypa/packaging.python.org/pull/1275/files#diff-6d2963726fc7c77dce51b2199325c2951882227016bf62bdd0340d84bdd6af11R106-R107 introduced a small merge conflict that needs addressing.

webknjaz commented:
> Is there some explanation of sigstore which could be linked? Even as someone who tries to stay on top of things and reads the python discourse, this one isn't familiar to me. Which suggests that most readers of this guide won't know what it is.

Not off the top of my head.

> Plus, with pypi phasing out gpg signatures, anything that sounds like "signing a package" probably deserves some extra explanation.

This is exactly why I wanted to have a ready-to-copy example. I've also been watching the development of the security changes around PyPI. GPG has been considered discouraged for like a decade, and Warehouse (the PyPI web app project) was working on the TUF among other things. Sigstore started emerging over the past few years and seems to be getting most traction. It is also targeting things outside the Python ecosystem, like containers.

I've seen some mentions of the possibility of integrating Sigstore but for now, the demo just shows how to publish signatures via GH releases.
I'm sure @di or @woodruffw could provide better insight, being more involved with it. My understanding is that they at least contributed to the Python implementation:
* https://blog.sigstore.dev/announcing-the-1-0-release-of-sigstore-python-4f5d718b468d/
* https://blog.trailofbits.com/2023/01/13/sigstore-python/

By the way, CPython itself started signing the python.org artifacts with Sigstore: https://www.python.org/download/sigstore/.
It's also been mentioned in the context of PyPI @ https://pbarker.dev/posts/2023-05-24/thoughts-on-pypi-pgp-and-sigstore/.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+      id-token: write
+
+    steps:
+     - name: Download all the dists
+       uses: actions/download-artifact@v3
+       with:
+         name: python-package-distributions
+         path: dist/
+     - name: Publish distribution 📦 to PyPI
+       uses: pypa/gh-action-pypi-publish@release/v1
```
Looks like this adds one-space indents for some reason.
```suggestion
    - name: Download all the dists
      uses: actions/download-artifact@v3
      with:
        name: python-package-distributions
        path: dist/
    - name: Publish distribution 📦 to PyPI
      uses: pypa/gh-action-pypi-publish@release/v1
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+    needs:
+    - publish-to-pypi
+    steps:
+     - name: Sign the dists with Sigstore
```
This also has one-space indents which is inconsistent with no-indent sequences in other places in this file.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+       # created the release above.
+       uses: softprops/action-gh-release@v1
+       with:
+         # dist/ contains the built packages, which smoketest-artifacts/
+         # contains the signatures and certificates.
```
Looks like I copied this from somewhere but it doesn't read well:
```suggestion
         # `dist/` contains the built packages, and the
         # sigstore-produced signatures and certificates.
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+      and upload them to GitHub Release
+    needs:
+    - publish-to-pypi
+    steps:
```
The first step should be the same call to `actions/download-artifact` as in the previous job — sigstore needs the files it'll sign to be on disk, in the locations mentioned as inputs.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+    name: >-
+      Sign the Python 🐍 distribution 📦 with Sigstore
+      and upload them to GitHub Release
+    needs:
+    - publish-to-pypi
```
The job needs these privileges. The target GHA VM is also mandatory for every job definition.
```suggestion
    needs:
    - publish-to-pypi
    runs-on: ubuntu-latest
    permissions:
      contents: write  # IMPORTANT: mandatory for making GitHub Releases
      id-token: write  # IMPORTANT: mandatory for trusted publishing & sigstore
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+      name: testpypi
+      url: https://test.pypi.org/p/<package-name>
+    permissions:
+      id-token: write
```
Let's show the comments as this is an educational document, after all:
```suggestion
      id-token: write  # IMPORTANT: mandatory for trusted publishing
```

Plus, it's important to record the context in which certain privileges are required.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+      # for <package-name>
+      url: https://pypi.org/p/<package-name>
+    permissions:
+      id-token: write
```
```suggestion
      id-token: write  # IMPORTANT: mandatory for trusted publishing
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      # Fill in your project (e.g. repository) name
```
In general, comments should be on the same line, if they fit. This helps prevent other lines from being injected in-between over time, which would make the comments drift away from their context.

webknjaz commented:
> I see the guide recommends using `softprops/action-gh-release`. I only looked into it minimally when I learned that github had deprecated their release artifact action, but some of that repo made our team a bit wary. (In particular, the github release stream vs the `v1` tag sends very mixed messages.) I've been tinkering with using the `gh` CLI to do this instead, which works something like
> 
> ```yaml
>       - name: Publish to GitHub Release
>         env:
>           GH_TOKEN: ${{ github.token }}
>           GH_REPO: ${{ github.repository }}
>         run: gh release upload "${{ github.ref_name }}" dist/*
> ```
> 
> Would it be interesting to put into the guide instead or in addition to the existing guidance?

I think it's a good idea for as long as it covers the same features being used. Although, I wouldn't use the env vars — they are likely unneeded. But if it's necessary, I'd just use CLI args. Also, this probably requires a trailing `**` to match files in subdirs.

webknjaz commented:
> > I already linked the [marketplace site ](https://github.com/marketplace/actions/gh-action-sigstore-python) of the `sigstore-python` GitHub Action (which originally is a command line tool) which contains all the documentation I have found.
> 
> Maybe I'm just out of the loop on this, but this is my first encounter with Sigstore. The result is that even with a link to the GitHub Action docs, I was missing all of the important context of "what is Sigstore?", "what value does signing with Sigstore provide my users?", "is this different from GPG signed packages?", etc.
> 
> I'm not sure exactly what linkage or explanation would have been best. Maybe a link going straight to https://docs.sigstore.dev/ is the right place?
> 
> Mostly, as an outsider providing feedback, I wanted to make sure there's no assumption that readers have any prior notion of what Sigstore is or how to use it. Telling users to use the action, but without the context to know what it does, naturally raises these questions.
> 
> I guess you could see this as a fault of the sigstore action docs? What matters most, IMO, is that as a reader is that I get relatively quickly to some place where I know whether or not I can or should omit that part of the workflow.

I fully agree with this. Sigstore is known among people who are directly in touch with security-related activities but often not to ordinary people.

Though, I'd probably link the root https://www.sigstore.dev site (or both root and docs). And add some context along the lines of "This is a replacement for GPG signatures, it allows signing the artifacts generated on GitHub. In this example, the signatures are uploaded to GitHub releases."

@chrysle @sirosen WDYT?

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+       # Confusingly, this action also supports updating releases, not
+       # just creating them. This is what we want here, since we've manually
+       # created the release above.
+       uses: softprops/action-gh-release@v1
```
@chrysle plz replace this action per https://github.com/pypa/packaging.python.org/pull/1261#issuecomment-1712995303 and @sirosen's suggestions.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,99 @@ should make GitHub run this workflow:
-PyPI, but only if the current commit is tagged. It is recommended
-you use the latest release tag; a tool like GitHub's dependabot can keep
-these updated regularly.
+.. collapse:: Load file
```
```suggestion
.. collapse:: Click here to display the entire GitHub Actions CI/CD workflow definition
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,99 @@ should make GitHub run this workflow:
-   :start-after: Actually publish to PyPI/TestPyPI
+   :start-at: publish-to-testpypi
+
+The whole CD workflow
```
This section looks weird with just one collapsed line. Can we add some paragraph at the beginning? Or think of some other structure?

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+    needs:
+    - build
+    runs-on: ubuntu-latest
+    environment:
```
```suggestion

    environment:
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,78 @@ jobs:
+    environment:
+      name: testpypi
+      url: https://test.pypi.org/p/<package-name>
+    permissions:
```
```suggestion

    permissions:
```

chrysle commented:
Thanks, I tried to apply your review comments best as possible!

> Though, I'd probably link the root https://www.sigstore.dev site (or both root and docs). And add some context along the lines of "This is a replacement for GPG signatures, it allows signing the artifacts generated on GitHub. In this example, the signatures are uploaded to GitHub releases."

Sure! I also linked the blog post announcing that PyPI has dropped support for PGP/GPG.

> I think it's a good idea for as long as it covers the same features being used. Although, I wouldn't use the env vars — they are likely unneeded. But if it's necessary, I'd just use CLI args. Also, this probably requires a trailing `**` to match files in subdirs.

Done! However, the env var seems to be necessary for the GitHub token, since I have not found a corresponding CLI arg.

sirosen commented:
> > I think it's a good idea for as long as it covers the same features being used. Although, I wouldn't use the env vars — they are likely unneeded. But if it's necessary, I'd just use CLI args. Also, this probably requires a trailing `**` to match files in subdirs.
> 
> Done! However, the env var seems to be necessary for the GitHub token, since I have not found a corresponding CLI arg.

Yeah, my understanding is that the env var is the simplest way of getting the token from the workflow context into the `gh` CLI. If there's a better way, that would be great, but this is what I have tested and verified for now.
Thanks for being so receptive to this suggested change!

> Though, I'd probably link the root https://www.sigstore.dev site (or both root and docs). And add some context along the lines of "This is a replacement for GPG signatures, it allows signing the artifacts generated on GitHub. In this example, the signatures are uploaded to GitHub releases."
> 
> [...] @sirosen WDYT?

That seems like a good summary explanation.
It wasn't that hard for me to get from `sigstore.dev` to the doc site, so any linkage in this family of sites is probably sufficient.

> By the way, CPython itself started signing the python.org artifacts with Sigstore: https://www.python.org/download/sigstore/.

I think this is *really* relevant, in that it shows that there's a lot of community buy-in for this signing solution.
Clearly I need to do some reading on the topic to catch-up! What do you think of including a mild reference to this, as it helps build readers' confidence in Sigstore if (like me) they hadn't heard of it?
e.g., "The following job signs the package with Sigstore, the same artifact signing system [used to sign CPython](https://www.python.org/download/sigstore/)."

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,88 @@ jobs:
+    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      # Fill in your project (e.g. repository) name for <package-name>
```
The term "repository" is quite overloaded. Let's not give the readers a change to assume it's a Git repo.. `pypi.org`, `test.pypi.org` and things like devpi deployments are also _[software]_ repositories.
```suggestion
      # Replace <package-name> below with your PyPI project name:
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,109 @@ should make GitHub run this workflow:
-So add this to the steps list:
+This additional job signs the distribution packages with `Sigstore`_,
+using the `sigstore/gh-action-sigstore-python GitHub Action`_,
+and then uploads them to GitHub Release.
```
```suggestion
and then uploads them to a GitHub Release.
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,109 @@ should make GitHub run this workflow:
 
-Finally, add the following steps at the end:
+   This is a replacement for GPG signatures, for which support has been
+   `removed <https://blog.pypi.org/posts/2023-05-23-removing-pgp/>`_ by PyPI.
```
how about
```suggestion
   `removed from PyPI <https://blog.pypi.org/posts/2023-05-23-removing-pgp/>`_.
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,88 @@ jobs:
+    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      # Fill in your project (e.g. repository) name for <package-name>
+      url: https://pypi.org/p/<package-name>
```
Oh, I realized that the earlier request to put it on the same line needs addressing too:
```suggestion
      url: https://pypi.org/p/<package-name>  # Replace <package-name> with your PyPI project name
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,88 @@ jobs:
+          ./dist/*.tar.gz
+          ./dist/*.whl
+    - name: Upload artifact signatures to GitHub Release
+      env:
+        GH_TOKEN: ${{ github.token }}
```
https://cli.github.com/manual/gh_help_environment says that the environment variable `GITHUB_TOKEN` is used for auth. I think GHA used to set it as an environment variable in the past but probably doesn't do so anymore. That's what confused me, I think. I'd still use that name because it's easier to grep for it.
```suggestion
      env:
        GITHUB_TOKEN: ${{ github.token }}
```

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
```diff
@@ -20,21 +20,88 @@ jobs:
+      # Upload to GitHub Release using the `gh` CLI.
+      # `dist/` contains the built packages, and the
+      # sigstore-produced signatures and certificates.
+      run: gh release upload "${{ github.ref_name }}" dist/** --repo "${{ github.repository }}"
```
I wonder if `--repo` is necessary... https://cli.github.com/manual/gh_release_upload says "Select another repository using the [HOST/]OWNER/REPO format" so the default is the current repository, I think.

chrysle replied:
Right, looks like it:

> GH_REPO: specify the GitHub repository in the "[HOST/]OWNER/REPO" format for commands that otherwise operate on a local repository.

https://cli.github.com/manual/gh_help_environment

webknjaz commented:
> What do you think of including a mild reference to this, as it helps build readers' confidence in Sigstore if (like me) they hadn't heard of it?
> e.g., "The following job signs the package with Sigstore, the same artifact signing system [used to sign CPython](https://www.python.org/download/sigstore/)."

That sounds reasonable! @chrysle would you include this too? Maybe add another commit with a co-author per https://hynek.me/til/easier-crediting-contributors-github/?

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,66 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+2. Fill in the name you wish to publish your new
+   :term:`PyPI project <Project>` under
+   (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),
+   the GitHub repository owner's name (org or user)
+   and repository name and the name of the release workflow file under
```
I think we should stick some commas here:
```suggestion
   the GitHub repository owner's name (org or user),
   and repository name, and the name of the release workflow file under
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,66 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+   the GitHub repository owner's name (org or user)
+   and repository name and the name of the release workflow file under
+   the ``.github/`` folder, see :ref:`workflow-definition`.
+   Finally add the name of the GitHub Actions environment
```
This is a bit confusing but the term is GitHub Environment. There's no "GitHub Actions environments" but the GitHub Environments can be specified in the GitHub Actions job definitions which is something different.
```suggestion
   Finally, add the name of the GitHub Environment
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,66 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+   (``pypi``) we're going set up under your repository.
+   Register the trusted publisher.
+3. Now, go to https://test.pypi.org/manage/account/publishing/ and repeat
+   the second step, but now enter ``testpypi`` as the name of the
```
Let's avoid the repetition of "now":
```suggestion
   the second step, but this time, enter ``testpypi`` as the name of the
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,66 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
+   Register the trusted publisher.
+3. Now, go to https://test.pypi.org/manage/account/publishing/ and repeat
+   the second step, but now enter ``testpypi`` as the name of the
+   GitHub Actions environment.
```
...and fix the terminology:
```suggestion
   GitHub Environment.
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -17,40 +19,66 @@ It will use the `pypa/gh-action-pypi-publish GitHub Action`_.
 
+   .. hint::
+
+      For security reasons, you should require manual approval
```
Many people probably wouldn't know what this is and where to set it up. We should either add this to the steps above or at least, link something like https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#deployment-protection-rules.

webknjaz commented on source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:
`actions/checkout@v3` should now be updated to `actions/checkout@v4` which got released recently.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,109 @@ should make GitHub run this workflow:
    :end-before: steps:
 
+This will also ensure that the PyPI publishing workflow is only triggered
+if the current commit is tagged. It is recommended you use the
+latest release tag.
```
I'm not sure what this last sentence means. Can we rephrase it to make it more illustrative?

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,109 @@ should make GitHub run this workflow:
-Finally, add the following steps at the end:
+   This is a replacement for GPG signatures, for which support has been
+   `removed <https://blog.pypi.org/posts/2023-05-23-removing-pgp/>`_ by PyPI.
+   However, this job is not mandatory for defining the workflow.
```
```suggestion
   However, this job is not mandatory for uploading to PyPI and can be omitted.
```

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,109 @@ should make GitHub run this workflow:
+   :start-after: path: dist/
    :end-before: steps:
 
+This will also ensure that the PyPI publishing workflow is only triggered
```
Let's also mention somewhere in this paragraph that it's important to set up the manual approval requirement for the `pypi` gh env.

webknjaz commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,109 @@ should make GitHub run this workflow:
    :language: yaml
-   :start-after: Actually publish to PyPI/TestPyPI
+   :start-at: publish-to-testpypi
+
```
```suggestion

.. tip::

   Requiring manual approvals in the `testpypi` GitHub Environment is typically unnecessary as it's designed to run on each commit to the main branch and is often used to indicate a healthy release publishing pipeline.


```

webknjaz reviewed (commented):
Hopefully, these comments are the last edits...

chrysle commented:
Addressed the comments!

> That sounds reasonable! @chrysle would you include this too? Maybe add another commit with a co-author per https://hynek.me/til/easier-crediting-contributors-github/?

To me too! Nice guide! Since I have a somewhat outdated `git` version on a somewhat outdated Ubuntu LTS not providing `--trailer`, I modified the script to work with a temporary commit template.

webknjaz commented:
> I have a somewhat outdated `git` version on a somewhat outdated Ubuntu LTS

These days, I tend to recommend new pythonistas to start by using Fedora as it's Python-friendly and ships modern software, being a creator-oriented distribution.

webknjaz reviewed (commented):
Let's fix these and get it merged.

webknjaz commented:
@chrysle I'm going to commit the suggestions myself now.

webknjaz commented:
I'm going to merge after https://github.com/pypa/packaging.python.org/pull/1261#discussion_r1324736220 is cleared, in case anybody wants to leave last reviews.

chrysle commented on source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:
```diff
@@ -67,69 +95,116 @@ should make GitHub run this workflow:
    :end-before: steps:
 
+This will also ensure that the PyPI publishing workflow is only triggered
+if the current commit is tagged. It is recommended that you commit using the
+latest release tag.
```
I haven't added this; see https://github.com/pypa/packaging.python.org/pull/1092#discussion_r929321386 for context. I see you have been struggling with that wording back then, too ;-). There is an unaccepted suggestion in the same discussion that could be added.

But I agree I understood it falsely at first, which is why the sentence is even more misleading now. 

I'm quite sure this means keeping the release tag of _gh-action-pypi-publish_ up-to-date (compare with the current version of the guide) . But since we have quite a lot of actions integrated now, I think we should either take this out or move it to a more obvious position.

webknjaz replied:
Oh, that didn't occur to me at all. Agreed, moving and making it clear that it's about the actions that are integrated is a good idea.

chrysle commented:
> These days, I tend to recommend new pythonistas to start by using Fedora as it's Python-friendly and ships modern software, being a creator-oriented distribution.

Thanks for suggesting! I shall try to overcome my obsession with Ubuntu (if they still haven't modernised anything with soon-to-come 24.04).

chrysle commented:
Done in the latest commit!

webknjaz reviewed (approved):
Fantastic job, @chrysle! Thank you for getting this across the finish line!

chrysle commented:
Many thanks to you all, especially @webknjaz, for your thorough reviews!
